miio/gateway.py

```diff
@@ -529,45 +581,295 @@ def set_light(self, color_name, brightness):
-                "Got an exception while fetching fw_ver: %s", ex, exc_info=True
-            )
-            return None
+            self.type = DeviceType(dev_info.type_id)
+        except ValueError:
+            self.type = DeviceType(-1)
 
     def __repr__(self):
         return "<Subdevice %s: %s fw: %s bat: %s>" % (
-            self.type,
+            self.device_type,
             self.sid,
             self.get_firmware_version(),
-            self.battery(),
+            self.get_battery(),
         )
 
+    @property
+    def device_type(self):
+        """Return the device type name."""
+        return self.type.name
+
+    @property
+    def battery(self):
+        """Return the battery level."""
+        return self._battery
+
+    @command()
+    def subdevice_send(self, command):
+        """Send a command/query to the subdevice"""
+        try:
+            return self._gw.send(command, [self.sid])
+        except Exception as ex:
+            raise GatewayException(
+                "Got an exception while sending command %s" % (command)
+            ) from ex
+            return None
+
+    @command()
+    def subdevice_send_arg(self, command, arguments):
+        """Send a command/query including arguments to the subdevice"""
+        try:
+            return self._gw.send(command, arguments, extra_parameters={"sid": self.sid})
+        except Exception as ex:
+            raise GatewayException(
+                "Got an exception while sending "
+                "command '%s' with arguments '%s'" % (command, str(arguments))
+            ) from ex
+
+    @command(click.argument("property"))
+    def get_subdevice_prop(self, property):
+        """Get the value of a property of the subdevice."""
+        try:
+            response = self._gw.send("get_device_prop", [self.sid, property])
+        except Exception as ex:
+            raise GatewayException(
+                "Got an exception while fetching property %s" % (property)
+            ) from ex
+
+        if not response:
+            raise GatewayException(
+                "Empty response while fetching property %s: %s" % (property, response)
+            )
+
+        return response
+
+    @command(click.argument("properties", nargs=-1))
+    def get_subdevice_prop_exp(self, properties):
+        """Get the value of a bunch of properties of the subdevice."""
+        try:
+            response = self._gw.send(
+                "get_device_prop_exp", [[self.sid] + list(properties)]
+            ).pop()
+        except Exception as ex:
+            raise GatewayException(
+                "Got an exception while fetching properties %s: %s" % (properties)
+            ) from ex
+
+        if len(list(properties)) != len(response):
+            raise GatewayException(
+                "unexpected result while fetching properties %s: %s"
+                % (properties, response)
+            )
+
+        for item in response:
+            if not item:
```

Comment:
I think this should be handled by the sub-device class as the interpretation of a non-true value can be device-specific, see my comment on the devices I tested against.

Reply:
I made it more specific, I now check if `item==''` which is the response if the property is not known by the gateway.

Reply:
Item can be empty also otherwise as shown in that output I gave, or I think that the values are what they are as the plug is not connnected. I don't exactly know where the plug is located, but if I find it tomorrow I'll plug it in and see what it returns after that :-)

Reply:
But if the device is not connected I think it is fine to raise an exception messaging that an empty response is received.
I just want to prevent that for instance the empty string '' is beeing devided by a number and raises an exception that is hard to understand (such as is done in the AqaraHT device update function where the raw response is devided by 100 to convert to degrees celsius).

Schould I remove the empty check?

Reply:
Yes, please remove this. You can add a check for the values prior the conversion (and raise an exception if the values are unexpected), or simply wrap the conversions inside try-catch block and raise a `GatewayException` with a friendly message when bad things happen.